File: CSES/Dynamic_Programming/Minimizing_Coins.cpp
#include <iostream>
#include <vector>
using namespace std;

int main(){
    int n,x;
    cin>>n>>x;
    vector<int> coins;
    vector<int> dp(x+1,1e9);
    for(int i=0;i<n;i++){
        int temp;
        cin>>temp;
        coins.push_back(temp);
    }
    dp[0]=0;
    for(int i=1;i<=x;i++){
        for(int j=0;j<n;j++){
            if(i-coins[j]>=0){
                dp[i]=min(dp[i],dp[i-coins[j]]+1);
            }
            
        }
    }
    if(dp[x]==1e9){
        cout<<-1;
    }
    else{
        cout<<dp[x];
    }
    
}

File: CSES/Dynamic_Programming/Minimizing_Coins.py
import sys
n,x=map(int,sys.readlines().split())
coins=list(map(int,sys.readlines().split()))
dp=[int(1e9)]*(x+1)
dp[0]=0
for i in range(x+1):
    for j in coins:
        if i-j >=0:
            dp[i]=min(dp[i],dp[i-j]+1)
if dp[x]==int(1e9):
    print(-1)
else:
    print(dp[x])
